WIP: Set up dynamic provisioning of persistent volumes on AWS (#423)
kinvolk/lokomotive: docs/configuration-reference/components/aws-ebs-csi-driver.md

# CSI Driver for Amazon EBS configuration reference for Lokomotive

## Contents

* [Introduction](#introduction)
* [Prerequisites](#prerequisites)
* [Configuration](#configuration)
* [Applying](#applying)

> * [Destroying](#destroying)

Reviewer: After https://github.com/kinvolk/lokomotive/pull/481, it should be `Deleting`, so we stay consistent.

## Introduction

> The [CSI Driver for Amazon EBS](https://github.com/kubernetes-sigs/aws-ebs-csi-driver)

Reviewer: I think we could make this description a bit more user friendly, to describe, that this component provides storage class for AWS, baked by Amazon EBS volumes.

We should also mention, that when cluster is destroyed, volumes must be cleaned up manually.

provides a CSI interface used by Container Orchestrators to manage the lifecycle
of Amazon EBS volumes.

## Prerequisites

* A Lokomotive cluster accessible via `kubectl` deployed on Packet.

## Configuration

> The CSI Driver component takes no confirguration arguments.

Reviewer: I think we should have a knob for this component, which enables/disables the default storage class. This way, the user can still use this component, but let's say have OpenEBS as a default storage option.

Author: How do configuration knobs work here? I added back `chartValuesTmpl`, hooked up the component struct, and added a boolean to `values.yaml`. I also wrapped the storageclass code in an `if` statement, but according to `lokoctl component render-manifest` it's still getting created even with my option set to false.

Reviewer: Hm, did you push the new code? I cannot see it, so it's hard to tell why is it like this. According to your description, everything seems correct.

Author: Pushed a WIP commit.

# CSI Driver for Amazon EBS configuration reference for Lokomotive

## Contents

* [Introduction](#introduction)
* [Prerequisites](#prerequisites)
* [Configuration](#configuration)
* [Attribute reference](#attribute-reference)
* [Applying](#applying)
* [Deleting](#deleting)

## Introduction

The [CSI Driver for Amazon EBS](https://github.com/kubernetes-sigs/aws-ebs-csi-driver)

> provides a CSI interface used by Container Orchestrators to manage the lifecycle

Reviewer 2: Could you please fix the caps here and everywhere else in this document?

Author: I'm not sure what you mean here. What should be changed?

Reviewer 3: I think he means this
```suggestion
provides a CSI interface used by container orchestrators to manage the lifecycle
```

> of Amazon EBS volumes. It provides a storage class for AWS, baked by Amazon EBS

Reviewer 2: ```suggestion
of Amazon EBS volumes. It provides a storage class for AWS, backed by Amazon EBS
```

volumes.

## Prerequisites

* A Lokomotive cluster accessible via `kubectl` deployed on Packet.

## Configuration

To run a cluster with the CSI Driver component, `set_csi_driver_iam_role` needs

> to be set to true in the `cluster` block of your lokocfg. The flag and the component

Reviewer 2: ```suggestion
to be set to `true` in the `cluster` block of your lokocfg. The flag and the component
```

should only be used for clusters deployed on AWS.

> CSI Driver component configuration example:

Reviewer 2: ```suggestion
Sample config:
```

> ```tf

Reviewer 2: ```suggestion
```hcl
```

# aws-ebs-csi-driver.lokocfg
component "aws-ebs-csi-driver" {
    enable_default_storage_class = true
}
```

## Attribute reference

Table of all the arguments accepted by the component.

Example:

| Argument                       | Description                                                  | Default      | Required |
|--------------------------------|--------------------------------------------------------------|:------------:|:--------:|
| `enable_default_storage_class` | Use the default storage class provided by the component      | true         | false    |

## Applying

> To apply the CSI Driver component:

Reviewer 2: ```suggestion
To apply the CSI Driver component, run the following command:
```

```bash
lokoctl component apply aws-ebs-csi-driver
```
By default, the CSI Driver pods run in the `kube-system` namespace.

## Deleting

> To destroy the component:

Reviewer 2: ```suggestion
To delete the component, run the following command:
```

```bash
lokoctl component delete aws-ebs-csi-driver
```

> When destroying the cluster or deleting the component, EBS volumes must be cleaned up
> manually.

Reviewer 2: ```suggestion
**WARNING: Before destroying a cluster or deleting the component, EBS volumes must be cleaned up
manually.** Failing to do so would result in EBS volumes being left behind.
```

> # CSI Driver for Amazon EBS configuration reference for Lokomotive

Reviewer 3: ```suggestion
# Amazon EBS CSI driver configuration reference for Lokomotive
```

* [Configuration](#configuration)
* [Attribute reference](#attribute-reference)
* [Applying](#applying)
* [Deleting](#deleting)

## Introduction

The [CSI Driver for Amazon EBS](https://github.com/kubernetes-sigs/aws-ebs-csi-driver)
provides a CSI interface used by Container Orchestrators to manage the lifecycle
of Amazon EBS volumes. It provides a storage class for AWS, backed by Amazon EBS
volumes.

## Prerequisites

* A Lokomotive cluster accessible via `kubectl` deployed on Packet.

## Configuration

To run a cluster with the CSI Driver component, `enable_csi` needs
to be set to `true` in the `cluster` block of your lokocfg. The flag and the component
should only be used for clusters deployed on AWS.

Sample config:

```hcl
# aws-ebs-csi-driver.lokocfg
component "aws-ebs-csi-driver" {
    enable_default_storage_class = true
}
```

## Attribute reference

Table of all the arguments accepted by the component.

Example:

| Argument                       | Description                                                  | Default      | Required |
|--------------------------------|--------------------------------------------------------------|:------------:|:--------:|

> | `enable_default_storage_class` | Use the default storage class provided by the component      | true         | false    |

Reviewer 3: I think this is more accurate.
```suggestion
| `enable_default_storage_class` | Use the storage class provided by the component as the default storage class      | true         | false    |
```
